### ISSUE TITLE
Mocking a step from a plugin does not work when testing vars/*.groovy 

### ISSUE THREAD

**Issue by Xaseron** [ExpediaGroup/jenkins-spock#34]
Expected Behavior
==============================
If i try to mock a step from a plugin when executing a step from vars/*.groovy it always executes the original step of the plugin. e.g. i'm trying to override `isUnix()` from the `workflow-basic-steps` plugin and it always executes the original step.

Actual Behavior
==============================
That the override would work.

Steps to Reproduce
==============================
`vars/exec.groovy`:
```groovy
void call(String command) {
    if (isUnix()) {
	    sh command
    } else{
	    bat command
    }
}
```
`ExecTest.groovy`:
```groovy
import com.homeaway.devtools.jenkins.testing.JenkinsPipelineSpecification

class ExecTest extends JenkinsPipelineSpecification {

    def Exec

    def setup() {
        Exec = loadPipelineScriptForTest('vars/exec.groovy')
    }

    def "Test on Windows"() {
        setup:
            getPipelineMock('isUnix')(_) >> { return false }

        when:
            Exec('ls')
        then:
            1 * getPipelineMock('isUnix') ()
            1 * getPipelineMock('bat') ('ls')
    }

    def "Test on Linux"() {
        setup:
            getPipelineMock('isUnix')(_) >> { return true }

        when:
            Exec('ls')
        then:
            1 * getPipelineMock('isUnix') ()
            1 * getPipelineMock('sh') ('ls')
    }
}
```

**Comment by awittha** [comment]
There is a typo/semantic problem that would prevent this from ever working:

```groovy
getPipelineMock('isUnix')(_) >> { return true }
```  

This is stubbing a call to `isUnix()` with one argument, but the code-under-test calls `isUnix()` with 0 arguments - so that would never stub the method call the mock actually received.


However, the real issue is a quirk of Spock: You cannot stub a mock in `setup` _and_ expect interactions with it in `then`. You must either:
 
1. Only expect interactions in `then`
2. Only stub interactions in `setup`
3. Combine expected interactions with stubbing in `then`

Spock's documentation on this: http://spockframework.org/spock/docs/1.0/interaction_based_testing.html#_combining_mocking_and_stubbing

Why didn't you get an error? IMO, Spock _should_ raise a "this isn't going to work the way you expect" error or at least a warning... But it doesn't.

Instead, the earlier stubbing is thrown out in favor of un-stubbed interaction expectations. This means that the mock pipeline step will return `null`, as all un-stubbed mocks do.

With `isUnix()` always returning `null`, and no explicit comparison operator provided in `exec.groovy`, Groovy will attempt to fit the `null` value into a `boolean` type. In Groovy, `null` is falsy, so the if statement always went to the `else` block.

In summary, the solutions are:
1. Fix the stubbing's parameters to match the actual method call
2. Fix the stubbing declaration so it is in the correct place for Spock
3. Consider a null-safe check on `isUnix()`'s results, as it may return _any_ kind of object - not necessarily a boolean - and Groovy will silently judge whether that value is true or false.

See the [`ExecSpec.groovy` test suite on the `issue-34` branch of this project](https://github.com/homeaway/jenkins-spock/blob/issue-34/examples/shared-library/test/ExecSpec.groovy) for working examples of the provided code, and additional unit tests that show what was going on.

**Comment by awittha** [comment]
FWIW, there is no Jenkins pipeline step execution (CpsScript / CpsThread / CpsFlowExecution), so it is not possible to actually execute pipeline steps from plugins during a `jenkins-spock` test-suite.

**Comment by awittha** [comment]
Please let me know if this adequately answers your question.

**Comment by Xaseron** [comment]
Thanks a lot for this in depth answer and the code examples.  :-)
